### PR TITLE
Added "bounded" parameter to optimizers, and using it when pertinent

### PR DESCRIPTION
**Pull request by fedeallocati**
The optimizers that don't have unbounded/bounded version give an assertion fail when calling them in that way

### CODE REVIEW

**Comment by fedeallocati** [comment]
Extensive tests compile

**Comment by fedeallocati** [comment]
Extensive tests run without error

File: src/limbo/opt/cmaes.hpp [hunk 1]
```diff
@@ -41,12 +43,12 @@ namespace limbo {
                 int irun, lambda = 0, countevals = 0;
                 char const* stop;
                 boundary_transformation_t boundaries;
-                double lowerBounds[] = {0.0}; // TODO  put this into params?
+                double lowerBounds[] = {0.0}; 
                 double upperBounds[] = {1.006309}; // Allows solution to be pretty close to 1
```
Comment (jbmouret) [inline src/limbo/opt/cmaes.hpp]: Why this magic number for the upper bounds?

**Comment by jbmouret** [comment]
- I think we should add default values when we have a preferred value. For example, for CMA-ES, we should default to true.

- I am wondering whether we need this as a template parameter or in the function argument. We could say that the bounded and unbounded version of an optimiser are two different optimisers (two types); and I do no think we need to change this at runtime.

What do you think?

File: src/limbo/opt/cmaes.hpp (hunk 1 above)
Comment (fedeallocati) [inline src/limbo/opt/cmaes.hpp]: Like the comment says, with it the solutions get really close to 1. If we put 1, the maximum that can be reached is a little lower than 1.

Comment (jbmouret) [inline src/limbo/opt/cmaes.hpp]: OK, but why "1.006309" and not "1.005" or something like this?

Comment (fedeallocati) [inline src/limbo/opt/cmaes.hpp]: Binary searched :p

Comment (jbmouret) [inline src/limbo/opt/cmaes.hpp]: Like this? https://en.wikipedia.org/wiki/Fast_inverse_square_root

On 27 Nov 2015, at 12:05, Federico Allocati <notifications@github.com> wrote:

> In src/limbo/opt/cmaes.hpp:
> 
> > @@ -41,12 +43,12 @@ namespace limbo {
> >                  int irun, lambda = 0, countevals = 0;
> >                  char const* stop;
> >                  boundary_transformation_t boundaries;
> > -                double lowerBounds[] = {0.0}; // TODO  put this into params?
> > +                double lowerBounds[] = {0.0}; 
> >                  double upperBounds[] = {1.006309}; // Allows solution to be pretty close to 1
> 
> Binary searched :p
> 
> �
> Reply to this email directly or view it on GitHub.
> 

� Jean-Baptiste Mouret
http://pages.isir.upmc.fr/~mouret/

Comment (fedeallocati) [inline src/limbo/opt/cmaes.hpp]: Exactly! I love that story haha

**Comment by fedeallocati** [comment]
> * I think we should add default values when we have a preferred value. For example, for CMA-ES, we should default to true.

I don't agree, because when we call them, we know if we want to use bounded or unbounded(e.g. in the acquisition optimization we always want to optimize in the unit square, and in the gp likelihood we always want unbounded search). Setting defaults in code that is templated, so you don't know which implementation you are receiving, doesn't seems like a good idea to me.

> *  I am wondering whether we need this as a template parameter or in the function argument. We could say that the bounded and unbounded version of an optimiser are two different optimisers (two types); and I do no think we need to change this at runtime.

We though about this, and the different ways to implement the bounded vs unbounded, and decided that having the parameter was the easier and cleaner way to do it. It also reinforces what I said to the other point. The BOptimizer doesn't care which inner optimization algorithm to use, as long as it allows for bounded search. I do think that having an assertion that fails at runtime if the optimizer doesn't support the corresponding bound mode isn't the best thing to do, but I think it's the best we can do without spending excessive time on it.

**Comment by jbmouret** [comment]
**OK** (but there are conflicts!).

**Comment by costashatz** [comment]
**OK for me** (fix the conflicts)

**Pull request merged**